### PR TITLE
[Config] [Connector-V2] Add Clickhouse and Assert Source/Sink Factory

### PR DESCRIPTION
**Pull request by Hisoka-X**
## Purpose of this pull request
1. Add Clickhouse and Assert Source/Sink Factory
2. Extend Option to support complex types

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  3. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  4. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by Hisoka-X** [comment]
@ashulin PTAL, specially about `Option`

File: docs/en/connector-v2/source/Clickhouse.md [hunk 1]
```diff
@@ -56,12 +55,6 @@ The query sql used to search data though Clickhouse server
 
 `ClickHouse` user password
 
-### schema [Config]
```
Comment (EricJoy2048) [inline docs/en/connector-v2/source/Clickhouse.md:59]: Why delete `schema`?

Comment (Hisoka-X) [reply docs/en/connector-v2/source/Clickhouse.md:59]: Clickhouse never support custom schema. It use schema projection.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/Options.java [hunk 2]
```diff
@@ -187,4 +207,18 @@ public Option<T> noDefaultValue() {
...
+
+        NestedTypedOptionBuilder(String key, TypeReference<T> typeReference, Option<?>[] acceptedOptions) {
+            super(key, typeReference);
+            this.acceptedOptions = acceptedOptions;
+        }
+
+        public Option<?>[] getAcceptedOptions() {
+            return acceptedOptions;
+        }
+    }
```
Comment (ashulin) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/Options.java]: `acceptedOptions` is not used.
I think it should change `OptionRule`

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/RequiredOption.java [hunk 3]
```diff
@@ -159,4 +159,45 @@ public String toString() {
             return String.format("Condition expression: %s, Required options: %s", expression, ExclusiveRequiredOptions.getOptionKeys(requiredOption));
         }
     }
+
+    /**
+     * All options must exist or not exist at the same time
+     */
+    class BothRequiredOptions implements RequiredOption {
```
Comment (ashulin) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/RequiredOption.java]:
```suggestion
    class BundledRequiredOptions implements RequiredOption {
```

Comment (ashulin) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/RequiredOption.java]: IMO, the word `Bundled` is more appropriate than `Both`.

File: seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/Config.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+
+    public static final Option<Map<String, Object>> FIELD_RULES_MAP = Options.key("field_rules_map").mapType(FIELD_NAME, FIELD_TYPE, FIELD_VALUE)
+        .noDefaultValue();
+
+    public static final Option<List<Map<String, Object>>> FIELD_RULES = Options.key("field_rules").listType(FIELD_RULES_MAP)
+        .noDefaultValue().withDescription("field rules for field validation");
+
+    public static final Option<Map<String, Object>> RULES = Options.key("rules").mapType(ROW_RULES, FIELD_RULES)
+        .noDefaultValue().withDescription("Rule definition of user's available data. Each rule represents one field validation or row num validation.");
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/Config.java]: IMO, this usage scenario can use `POJO` to represent `rules`

File: seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/Config.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+package org.apache.seatunnel.connectors.seatunnel.assertion.sink;
+
+import org.apache.seatunnel.api.configuration.Option;
+import org.apache.seatunnel.api.configuration.Options;
+import org.apache.seatunnel.connectors.seatunnel.assertion.rule.AssertFieldRule;
+
+import java.util.List;
+import java.util.Map;
+
+public class Config {
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/Config.java]:
```suggestion
public class AssertConfig {
```

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/RequiredOption.java (hunk 3 above)
Comment (TyrantLucifer) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/RequiredOption.java]: +1

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/config/Config.java [hunk 6]
```diff
@@ -17,80 +17,100 @@
...
 
+import org.apache.seatunnel.api.configuration.Option;
+import org.apache.seatunnel.api.configuration.Options;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+@SuppressWarnings("checkstyle:MagicNumber")
 public class Config {
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/config/Config.java]:
```suggestion
public class ClickhouseConfig {
```

**Comment by Hisoka-X** [comment]
@ashulin @TyrantLucifer @EricJoy2048 Already changed, PTAL. Thanks!

File: seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/AssertSinkFactory.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+package org.apache.seatunnel.connectors.seatunnel.assertion.sink;
+
+import static org.apache.seatunnel.connectors.seatunnel.assertion.sink.AssertConfig.RULES;
+
+import org.apache.seatunnel.api.configuration.util.OptionRule;
+import org.apache.seatunnel.api.table.factory.TableSinkFactory;
+
+import com.google.auto.service.AutoService;
+
+@AutoService(TableSinkFactory.class)
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/sink/AssertSinkFactory.java]:
```suggestion
@AutoService(Factory.class)
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.SPLIT_MODE;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.TABLE;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.USERNAME;
+
+import org.apache.seatunnel.api.configuration.util.OptionRule;
+import org.apache.seatunnel.api.table.factory.TableSinkFactory;
+
+import com.google.auto.service.AutoService;
+
+@AutoService(TableSinkFactory.class)
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java]:
```suggestion
@AutoService(Factory.class)
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/file/ClickhouseFileSinkFactory.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.SHARDING_KEY;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.TABLE;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.USERNAME;
+
+import org.apache.seatunnel.api.configuration.util.OptionRule;
+import org.apache.seatunnel.api.table.factory.TableSinkFactory;
+
+import com.google.auto.service.AutoService;
+
+@AutoService(TableSinkFactory.class)
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/file/ClickhouseFileSinkFactory.java]:
```suggestion
@AutoService(Factory.class)
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceFactory.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.PASSWORD;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.SQL;
+import static org.apache.seatunnel.connectors.seatunnel.clickhouse.config.ClickhouseConfig.USERNAME;
+
+import org.apache.seatunnel.api.configuration.util.OptionRule;
+import org.apache.seatunnel.api.table.factory.TableSourceFactory;
+
+import com.google.auto.service.AutoService;
+
+@AutoService(TableSourceFactory.class)
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceFactory.java]:
```suggestion
@AutoService(Factory.class)
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+public class ClickhouseSinkFactory implements TableSinkFactory {
+    @Override
+    public String factoryIdentifier() {
+        return "ClickhouseSink";
+    }
+
+    @Override
+    public OptionRule optionRule() {
+        return OptionRule.builder().required(HOST, DATABASE, TABLE)
+            .optional(CLICKHOUSE_PREFIX, BULK_SIZE, SPLIT_MODE, FIELDS, SHARDING_KEY)
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java:47]: From the document https://seatunnel.apache.org/docs/2.3.0-beta/connector-v2/sink/Clickhouse I see `fields ` is required.

**Review by EricJoy2048 — APPROVED** [review]

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java (hunk 11 above)
Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/ClickhouseSinkFactory.java:47]: It should be required

**Comment by Hisoka-X** [comment]
@ashulin @TyrantLucifer PTAL again, Thanks!

**Review by TyrantLucifer — APPROVED** [review]
+1

**Pull request merged**